### ISSUE TITLE
Unsound assignment to function type taken from an interface

### ISSUE THREAD

**Issue by ealmansi-uc** [microsoft/TypeScript#49378]
# Bug Report

### 🔎 Search Terms

typescript, interface, assignment, unsound, covariance, contravariance

### 🕗 Version & Regression Information

- This is the behavior in every version I tried.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?ts=4.7.2#code/JYOwLgpgTgZghgYwgAgBrIN4ChnIB4D8AXMgM5hSgDmA3FgL5aiSyIoCamO+J5lItBk3DR4SZAEkuuGAAoA+iVQBKEgDcA9sAAmdRljABPAA4oAYgEZkAXmQKlymwD5kmnXSOnkZgEw3JANoARDBBALp0BiYoAOL+9sjsjtYubrpYCBog5MhUJHG2shj4yPTJLsV4AHQUwAC2so70kQD0LcgA8gDSJADCGnXGwAA20MjQUBpQJAAq0cgA5ACuINoQMKAQ2gvIwKTIIBpgyHCkpMBUIHAARqPIYBr38wt81As1pLI+AMw+PspYNrITLZY4wCwkSz+QSA9oAOSOnR6B0eEymVQyWRyMB8kL8thhQIASiswPUUGjpshenAQIdjlAIHBtMhjJNTFAyRB9hoYMgVmsNiAtnZGczqItanUFsoMTiimU6EA)

### 💻 Code

```ts
interface X {
  x?: string;
}
interface Y {
  x: string;
}
interface I {
  f(_: X): void;
}

type F1 = (_: X) => void;
type F2 = I["f"];

type G = (_: Y) => void;
const g: G = ({ x }) => { x.trim() };

// OK: Compiler error: Type 'undefined' is not assignable to type 'string'.ts(2322)
// const f1: F1 = g;

// Not OK: no error.
const f2: F2 = g;

// Runtime error: Cannot read properties of undefined (reading 'trim').
f2({});
```

### 🙁 Actual behavior

Typescript compiles the code, which then throws a runtime error when executed.

### 🙂 Expected behavior

Declaration and assignment to `f2` should be a compiler error, since type G is not compatible with F2.
The error is detected correctly for `f1`, which is interesting because F1 and F2 are equivalent.

**Comment by Roukanken47** [comment]
Cleaning it up a bit:

```js
// input types to play with
// replacing OriginalArgment with commented line makes assigment to a fail too, 
// despite property b being irrelevant

// type OriginalArgument = { a: string; b: string; };
type OriginalArgument = { a: string; };
type ChangedArgument = { a: string; c: string };

interface Interface {
  func(_: OriginalArgument): void;
}

type A = Interface["func"];
type B = (_: OriginalArgument) => void;

const implementation = (_: ChangedArgument) => {};

const a: A = implementation;
const b: B = implementation;
```

As long as `ChangedArgument` is assignable to `OriginalArgument` then assignment to `a` won't throw error, despite the fact that the needed check is "`OriginalArgument` is assignable to `ChangedArgument`". Assigment to `b` behaves properly.

(Note: if `func` is defined as property with lambda function, everything works properly.)

**Comment by IllusionMH** [comment]
`strictFunctionTypes` won't apply to **methods** in the interface but will for properties that are functions.

Changing interface to code below should provide expected error.
```ts
interface I {
  f: (_: X) => void;
}
```

Not sure duplicate of which issue this one is, but #47988 has similar behavior event if code isn't one to one match.

**Comment by ealmansi-uc** [comment]
@IllusionMH thanks for your input - I guess that settles the topic

Pretty much the same behaviour I described is mentioned here: https://www.typescriptlang.org/tsconfig#strictFunctionTypes

> During development of this feature, we discovered a large number of inherently unsafe class hierarchies, including some in the DOM. Because of this, the setting only applies to functions written in function syntax, not to those in method syntax:

```
type Methodish = {
  func(x: string | number): void;
};
 
function fn(x: string) {
  console.log("Hello, " + x.toLowerCase());
}
 
// Ultimately an unsafe assignment, but not detected
const m: Methodish = {
  func: fn,
};
m.func(10);
```

**Comment by fatcerberus** [comment]
Yes, this behavior is by design.  See here for rationale:
https://github.com/microsoft/TypeScript/wiki/FAQ#why-are-function-parameters-bivariant

tl;dr: Methods are bivariant so that array types remain covariant.
